### PR TITLE
Fix closure untupling/retupling

### LINKED ISSUE #240: Incorrect verification results for function/closure dyn traits
The following fails to verify:

```rust
fn foo(fun: &mut dyn FnMut(i32) -> i32) -> i32 {
    fun(1)
}

fn main() {
    let r = foo(&mut move |z : i32| { z } );
    assert!(r == 1); // should succeed
}
```

In the generated CBMC-C code, in some places we pass the closure argument as a tuple, but in others we expect the flattened int.

### PR DESCRIPTION
**Pull request by avanhatt**
### Description of changes: 

RMC currently fails to verify some nested boxed closures, such as this:
```rust
    let f: Box<Box<dyn FnOnce(f32, i32)>> = Box::new(Box::new(|x, y| { assert!(x == 1.0); assert!(y == 2) }));
    f(1.0, 2);
```

Here is an example of the code we now generate (with renaming for clarity):
```
struct Unit call_once_vtable_shim(struct [boxed_closure.rs:8:49: 11:6] *var_1, float spread_2, int spread_3)
{
  struct (f32, i32) var_2={ .0=spread_2, .1=spread_3 }; // new: retuple since the body requires it
  struct Unit var_0;

bb0:
  ;
  boxed_closure_main(*var_1, var_2.0, var_2.1);

bb1:
  ;
  return var_0;
}
```

The problem is that we were checking whether to untuple closure arguments based on whether the function's type kind was a `Closure`, but this misses some cases (where the trait points to a defined function, or with vtable shims) where the type is a `FnDef`. The right way to check for untupling is via a check to the function's ABI. However, in that case, we also need to _retuple_ arguments based on the MIR's `spread_arg` flag. See [this Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Determine.20untupled.20closure.20args.20from.20Instance.3F) for more discussion.

The trickiness here is that we need to insert both the tuple and its individual components into the symbol table, and then write the components out to the tuple at the start of a function. To do this, I added a new naming convention for the untupled args, avoiding naming conflicts based on the local index. Local index math determines what gets retupled when necessary.

### Resolved issues:

Resolves #240

### Call-outs:

`ignore_var_ty` is mystifying, and there are some intrinsic cases where skipping `FnDef` there elides the environment argument of a call through a tuple. I don't like this, and we should probably eventually remove `ignore_var_ty` in favor of elides arguments only for specific, documented reasons (like the being Zero Sized Types that rust elides). 

### Testing:

* How is this change tested?

Existing tests, including a `fixme` added in my last PR.

* Is this a refactor change?

Somewhat.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by adpaco-aws** [comment]
Hi @avanhatt , can you please add the example to the description?

File: compiler/rustc_codegen_llvm/src/gotoc/mod.rs [hunk 1]
```diff
@@ -156,6 +161,84 @@ impl<'tcx> GotocCtx<'tcx> {
...
 
+    /// MIR functions have a `spread_arg` field that specifies whether the
+    /// final argument to the function is "spread" at the LLVM/codegen level
+    /// from a tuple into its individual components. (Used for the "rust-
+    /// call" ABI, necessary because dynamic trait closure cannot have an
+    /// argument list in MIR that is both generic and variadic, so Rust
+    /// allows a generic tuple).
+    ///
+    /// If `spread_arg` is Some, then the wrapped value is the local that is
+    /// to be "spead"/untupled. However, the MIR function body itself expects
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/mod.rs:172]:
```suggestion
    /// to be "spread"/untupled. However, the MIR function body itself expects
```

File: src/test/cbmc/DynTrait/dyn_fn_param.rs [hunk 2]
```diff
@@ -1,13 +1,6 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0 OR MIT
 
-//rmc-flags: --no-memory-safety-checks
-
-// We use `--no-memory-safety-checks` in this test to avoid getting
-// a verification failure:
-// [pointer_dereference.14] invalid function pointer: FAILURE
-// Tracking issue: https://github.com/model-checking/rmc/issues/307
-
```
Comment (adpaco-aws) [inline src/test/cbmc/DynTrait/dyn_fn_param.rs:10]: Nice!

File: compiler/rustc_codegen_llvm/src/gotoc/metadata.rs [hunk 3]
```diff
@@ -283,30 +307,41 @@ impl<'tcx> GotocCtx<'tcx> {
...
 
     pub fn fn_sig_of_instance(&self, instance: Instance<'tcx>) -> ty::PolyFnSig<'tcx> {
         let fntyp = instance.ty(self.tcx, ty::ParamEnv::reveal_all());
         self.monomorphize(match fntyp.kind() {
             ty::Closure(def_id, subst) => self.closure_sig(*def_id, subst),
-            _ => fntyp.fn_sig(self.tcx),
+            ty::FnPtr(..) | ty::FnDef(..) => {
+                let sig = fntyp.fn_sig(self.tcx);
+                // Some virtual calls through a vtable may actually be closures
+                // that also need the arguments uptupled, even though the kind of
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/metadata.rs:337]: Does "uptupled" have a typo?

File: compiler/rustc_codegen_llvm/src/gotoc/mod.rs [hunk 4]
```diff
@@ -146,6 +146,11 @@ impl<'tcx> GotocCtx<'tcx> {
             let loc = self.codegen_span2(&ldata.source_info.span);
             let sym =
                 Symbol::variable(name, base_name, t, self.codegen_span2(&ldata.source_info.span));
+            if self.symbol_table.contains(&sym.name) {
+                debug!("Skipping already-added local: {:?}", sym.name);
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/mod.rs:150]: Why do we need this check?

File: compiler/rustc_codegen_llvm/src/gotoc/statement.rs [hunk 5]
```diff
@@ -195,31 +195,50 @@ impl<'tcx> GotocCtx<'tcx> {
...
+        //     0. a struct representing the environment
+        //     1. a tuple containing the parameters
+        //
+        // However, for some reason, Rust decides to generate a function which still
+        // takes the first argument as the environment struct, but the tuple of parameters
+        // are flattened as subsequent parameters.
+        // Therefore, we have to project out the corresponding fields when we detect
+        // an invocation of a closure.
+        //
+        // Note: in some cases, the enviroment struct has type FnDef, so we skip it in
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/statement.rs:219]:
```suggestion
        // Note: In some cases, the enviroment struct has type FnDef, so we skip it in
```

File: compiler/rustc_codegen_llvm/src/gotoc/utils.rs [hunk 6]
```diff
@@ -25,6 +25,14 @@ impl<'tcx> GotocCtx<'tcx> {
         }
     }
 
+    // Special naming conventions for parameters that are spread from a tuple
+    // into its indivdual components at the LLVM level, see comment at
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/utils.rs:29]:
```suggestion
    // into its individual components at the LLVM level, see comment at
```

File: compiler/rustc_codegen_llvm/src/gotoc/utils.rs [hunk 7]
```diff
@@ -25,6 +25,14 @@ impl<'tcx> GotocCtx<'tcx> {
...
+    // compiler/rustc_codegen_llvm/src/gotoc/mod.rs:codegen_function_prelude
+    pub fn codegen_spread_arg_name(&self, l: &Local) -> String {
+        let fname = self.current_fn().name();
+        format!("{}::1::spread{:?}", fname, l)
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/utils.rs:33]: Do we need to hard-code this?

**Review by adpaco-aws — COMMENTED** [review]
Great work! A few minor comments and questions.

File: compiler/rustc_codegen_llvm/src/gotoc/mod.rs (hunk 4 above)
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/mod.rs]: The spread arg is added in the prelude, so we shouldn't re-add it to the table here. But I added a different, more explicit check and comment instead, thanks!

File: compiler/rustc_codegen_llvm/src/gotoc/utils.rs (hunk 7 above)
Comment (avanhatt) [reply compiler/rustc_codegen_llvm/src/gotoc/utils.rs]: I think so, the name `spread` is not special, we just need to name it something different from `var` and I chose this to try and help convey the purpose.

File: compiler/rustc_codegen_llvm/src/gotoc/mod.rs [hunk 8]
```diff
@@ -135,10 +135,12 @@ impl<'tcx> GotocCtx<'tcx> {
         let mir = self.current_fn().mir();
         let ldecls = mir.local_decls();
         ldecls.indices().for_each(|lc| {
-            let base_name = match self.find_debug_info(&lc) {
-                None => format!("var_{}", lc.index()),
-                Some(info) => format!("{}", info.name),
-            };
+            if Some(lc) == mir.spread_arg {
+                // We have already added this local in the functin prelude, so
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/mod.rs:139]:
```suggestion
                // We have already added this local in the function prelude, so
```

**Review by adpaco-aws — APPROVED** [review]
LGTM! Do you mind updating manually and fixing the typo below?

**Pull request merged**